js/util/settings/settings.js

```diff
@@ -1,17 +1,44 @@
 var settings = {
   filePath: (process.type === 'renderer' ? window.globalArgs['user-data-path'] : userDataPath) + (process.platform === 'win32' ? '\\' : '/') + 'settings.json',
+  fileWritePromise: null,
   list: {},
   onChangeCallbacks: [],
   save: function (cb) {
-    fs.writeFile(settings.filePath, JSON.stringify(settings.list), function (e) {
+    /*
+    Writing to the settings file from multiple places simultaneously causes data corruption, so to avoid that:
+    * We forward data from the renderer process to the main process, and only write from there
+    * In the main process, we put multiple save requests in a queue (by chaining them to a promise) so they execute individually
+    * https://github.com/minbrowser/min/issues/1520
+    */
+
+    if (process.type === 'renderer') {
+      ipc.send('receiveSettingsData', settings.list)
       if (cb) {
         cb()
       }
-    })
-    if (process.type === 'renderer') {
-      ipc.send('receiveSettingsData', settings.list)
-    } else if (process.type === 'browser' && mainWindow) {
-      mainWindow.webContents.send('receiveSettingsData', settings.list)
+    }
+
+    if (process.type === 'browser') {
+      /* eslint-disable no-inner-declarations */
+      function writeFileInner () {
+        settings.fileWritePromise = fs.promises.writeFile(settings.filePath, JSON.stringify(settings.list)).then(function (e) {
+          settings.fileWritePromise = null
+          if (cb) {
+            cb()
+          }
+        })
+      }
+      /* eslint-enable no-inner-declarations */
+
+      if (settings.fileWritePromise) {
+        settings.fileWritePromise.then(writeFileInner)
+      } else {
+        writeFileInner()
+      }
```

Comment:
I think the logic works fine but just proposing a minor refactor to make the promise chain (perhaps?) a little bit more readable.  
```suggestion
      /* eslint-disable no-inner-declarations */
      function newFileWrite () {
        return fs.promises.writeFile(settings.filePath, JSON.stringify(settings.list)).then(function (e) {
          if (cb) {
            cb()
          }
        })
      }

      function ongoingFileWrite () {
        return settings.fileWritePromise ? settings.fileWritePromise : Promise.resolve()
      }
      /* eslint-enable no-inner-declarations */

      // eslint-disable-next-line no-return-assign
      settings.fileWritePromise = ongoingFileWrite().then(newFileWrite).then(() => settings.fileWritePromise = null)
```